feat: Indicate modified buffers

chomosuke opened the pull request:
closes #1577.

New options and their docs are already completed, I'm opened to feedback on those. 

TODOs:
- [x] highlight_modified
- [x] modified_placement
  - [x] after
  - [x] before
  - [x] signcolumn
- [x] show_on_dirs
- [x] show_on_open_dirs

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -324,6 +327,12 @@ Subsequent calls to setup will replace the previous configuration.
         show_on_open_dirs = true,
         timeout = 400,
       },
+      modified = {
```
Please ignore the above; that's inconsistent and magic and loses functionality.

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -240,6 +241,8 @@ Subsequent calls to setup will replace the previous configuration.
           git_placement = "before",
           padding = " ",
           symlink_arrow = " ➛ ",
+          modified = "[+]",
```
We should use a single character `glyphs.modified`

There is code in the renderer that assumes icons are single characters.

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -542,6 +551,7 @@ Configuration options for the system open command.
 
 *nvim-tree.diagnostics*
 Show LSP and COC diagnostics in the signcolumn
+Note that the modified sign will take precedence over the diagnostics signs.
```
:+1: that is very friendly.

alex-courtis reviewed (changes requested):
Looking good so far.

## OK

defaults

```lua
    modified = {
      enable = true,
      show_on_dirs = true,
      show_on_open_dirs = false,
      debounce_delay = 50,
    },
```

```lua
    update_focused_file = {
      enable = true,
      debounce_delay = 15,
```

Manual refresh

:NvimTreeToggle

:NvimTreeClose :NvimTreeOpen

Change directory to a different sibling then back

## NOT OK

does not show on opened directory
```lua
    modified = {
      enable = true,
      show_on_dirs = true,
      show_on_open_dirs = false,
      debounce_delay = 50,
    },
```

does not show on any directory
```lua
    modified = {
      enable = true,
      show_on_dirs = true,
      show_on_open_dirs = true,
      debounce_delay = 50,
    },
```

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -859,6 +904,11 @@ UI rendering setup
 
+            *nvim-tree.renderer.icons.glyphs.modified*
+            Icon to display for modified files. Can be multiple characters if
+            |renderer.icons.modified_placement| isn't `"signcolumn"`.
```
That's quite nice. We will need to do some extensive testing of `get_padding` as that is quite... fragile.

We could allow multiple only for `"after"` to avoid that. It would be consistent with vim's `[+]` after which I imagine is the use case.

alex-courtis commented on lua/nvim-tree/modified.lua:
```diff
@@ -0,0 +1,34 @@
+  end
+end
+
+---@param node table TODO: we should probably have a Node type
```
We do: https://github.com/nvim-tree/nvim-tree.lua/blob/89c79cb33bc4867ccf9ee4810697da76ac59fdba/lua/nvim-tree/explorer/node-builders.lua#L13

That's one of the reasons I want to make a Node class...

alex-courtis replied:
Oh I see, you meant Node class, not a type field.

In the future. I'd like to have subclasses for file, link etc. as well as a subclass for `TreeExplorer`

alex-courtis commented on lua/nvim-tree.lua:
```diff
@@ -633,6 +648,12 @@ local DEFAULT_OPTS = { -- BEGIN_DEFAULT_OPTS
+    enable = false,
+    show_on_dirs = true,
+    show_on_open_dirs = false,
+    debounce_delay = 50,
```
Could you please implement #1836 in this PR.

legacy.lua `refactored` can be used to silently migrate `update_focused_file.debounce_delay`

chomosuke replied:
Sure, I'll take it look.

chomosuke replied:
Actually do you mind if I remove `modified.debounce_delay` for now and do #1836 in another PR? I'm not too comfortable making this already pretty big PR any bigger.

alex-courtis replied:
> Actually do you mind if I remove `modified.debounce_delay` for now and do #1836 in another PR? I'm not too comfortable making this already pretty big PR any bigger.

No worries. That is sensible.

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -859,6 +904,11 @@ UI rendering setup
 
+            *nvim-tree.renderer.icons.glyphs.modified*
+            Icon to display for modified files. Can be multiple characters if
+            |renderer.icons.modified_placement| isn't `"signcolumn"`.
```
That's quite nice. I dont't think `get_padding` will handle multiple characters before, so we should restrict multiple characters to "after" only.

It would be consistent with vim's [+] after which I imagine is the use case.

chomosuke replied:
Hmmmm, with my testing `get_padding` only assume `renderer.indent_markers.icons` is a single character? everything else seems to be fine with multiple character.
This is on `master`:
![image](https://user-images.githubusercontent.com/38484873/209271149-e42b8606-3d29-4048-bc6a-ca71929029ff.png)
I did see some code assuming ` renderer.icons.padding` is a single character though. Though my next commit will make that no longer the case due to some refactoring.

chomosuke replied:
Update: upon further testing, it looks like we assume all icons that could show up in the sign column are at most 2 characters. This is because `vim.fn.sign_define()` doesn't allow more than two characters. We can fix this in another PR by only calling `vim.fn.sign_define()` when placement is actually `"signcolumn"`.

chomosuke commented:
Test cases:

- [ ] padding
  - [ ] empty string
  - [ ] long string
- [ ] git, modified placement:
  - [x] signcolumn, signcolumn
    - [x] diagnostics & modified precede git
    - [x] modified precede diagnostics
  - [ ] before, before
  - [ ] before, after
  - [ ] after, before
  - [ ] after, after
- [ ] highlight_modified
  - [ ] icon
    - [ ] precedence
  - [ ] name
    - [ ] precedence
  - [ ] change highlight group color

alex-courtis commented:
We could add to base16 when we are done https://github.com/echasnovski/mini.nvim/pull/181

alex-courtis commented on lua/nvim-tree/renderer/builder.lua:
```diff
@@ -91,46 +97,48 @@ local function get_folder_name(node)
-  if not git_icons_and_hl_groups then
+---@class HighlightedString
+---@field str string
+---@field hl string|nil
```
Thanks for starting this documentation pattern. We should use it more.

`sumneko_lua` gives me a warning however it does that for other lua code. I might look at putting a suppression in.

![20221230_171338 1052x253](https://user-images.githubusercontent.com/1505378/210040345-ac0b97d4-bf7b-4297-976a-c6feb0553583.png)

chomosuke replied:
hmmm interesting, it doesn't give me any warnings. But maybe putting in suppression in `.luarc.json` is a good idea for this one.

alex-courtis replied:
The problem was on my end... I was putting the current directory in the workspace twice.

I've actually managed to remove all the suppressions: https://github.com/nvim-tree/nvim-tree.lua/commit/e322fbb80b0be5e885268c91ab29cbfe0143cd5b

alex-courtis reviewed (changes requested):
Looking good, apologies for the delay.

Tested some with the following general cases:
- git
- diagnostics
- symlinks
- indent markers

Testing all permutations and combinations is not feasible however sampled the following:
- [X] `modified.enabled`
- [X] `modified.show_on_dirs`
- [X] `modified.show_on_open_dirs`
- [X] `highlight_modified = "none"`
- [X] `highlight_modified = "icon"`
- [X] `highlight_modified = "name"`
- [X] `highlight_modified = "all"`
- [X] `highlight_modified = "bleh"`
- [X] `glyphs.modified = ""`
- [X] `glyphs.modified = "X"`
- [X] `icons.modified_placement = "after"`
- [X] `icons.modified_placement = "before"`
- [X] `icons.modified_placement = "signcolumn"`
- [X] `icons.show.modified`

"Failing" but good enough as they are consistent
- [X] `glyphs.modified = "XY"` shows both, missing, same as git and file icons, good enough
- [X] `icons.modified_placement = "bleh"` missing, same as git, good enough

TODO
- [ ] fix extra folder padding https://github.com/nvim-tree/nvim-tree.lua/pull/1835#discussion_r1059246240

chomosuke commented on doc/nvim-tree-lua.txt:
```diff
@@ -324,6 +328,11 @@ Subsequent calls to setup will replace the previous configuration.
+      modified = {
+        enable = false,
+        show_on_dirs = true,
+        show_on_open_dirs = false,
```
I actually wanted to change git & diagnostics to the same as modified because I think that'd be the default that most people want, but didn't cause i thought that'd be breaking.

Idk what is this plugin's stance on changes in default though, if you think it's alright I can make that change in this PR.

In the mean time I'll change modified to be consistent with git for now :).

alex-courtis replied:
Yeah, I'd like to change those defaults, they make more sense. That's what I use.

Users do get upset when we make changes like that...

I am open to trying it again: how about we make a separate PR for those sensible defaults that we can revert if there is upset.

gegoune replied:
I have this idea in my head for a while now. Wanted to open an issue but didn't find time to do investigation deep enough to justify it. 
I would like to discuss future of the project and breaking changes moving forward. I wanted to argue in favour of releasing pre-release version with multiple breaking changes: removed/unified functionality, reorganised configuration options, general clean up. I think project could greatly benefit from refactor without worrying about breaking changes. It would then be released as version 2 for example. Would there be any appetite for that at all?

alex-courtis replied:
> reorganised configuration options, general clean up

Yes. `git.ignore` not being in filters. `renderer` too large. `diagnostics.icons` not in `icons.glyphs`. Some root options not in `view` or their own section.

> I would like to discuss future of the project and breaking changes moving forward.

We can break things in a controlled manner that will result in a better experience. We are also good at refactoring options, although we don't use it much.

> I wanted to argue in favour of releasing pre-release version with multiple breaking changes

> It would then be released as version 2 for example.

Many lua plugins and packer are getting very good at handling versions. I think it is time.

`master` could remain as "version 1" to avoid interruptions to users, however we could add `devel`, `weekly` etc. along with `stable`, `stable-2`, `stable-2.1`, `stable-2.1.3` etc. I like SEMVER.

alex-courtis replied:
I did discover an interesting pattern with lualine:

User is prompted to run `:LualineNotices` when there is a configuration problem.

```
theme(base16): nvim-base16 is not currently present in your runtimepath, make sure it is properly installed, fallback to default colors.
```

gegoune replied:
That's great to hear. Let's open a dedicated issue to discuss it further.

alex-courtis replied:
https://github.com/nvim-tree/nvim-tree-management

alex-courtis commented on lua/nvim-tree/renderer/components/padding.lua:
```diff
@@ -68,6 +68,12 @@ local function get_padding_arrows(node, indent)
   end
 end
 
+---@param depth integer
```
Works nicely.

alex-courtis reviewed (changes requested):
Last one:

All defaults but `modified = "[+]",`
```
Error executing vim.schedule lua callback: ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:359: Vim(lua):E5108: Error executing lua Vim:E239: Invalid sign text: [+]
stack traceback:
        [C]: in function 'sign_define'
        ...start/dev/lua/nvim-tree/renderer/components/modified.lua:19: in function 'setup_signs'
        ...start/dev/lua/nvim-tree/renderer/components/modified.lua:36: in function 'setup'
        ...te/pack/packer/start/dev/lua/nvim-tree/renderer/init.lua:113: in function 'setup'
        /tmp/nvt-dev/site/pack/packer/start/dev/lua/nvim-tree.lua:828: in function 'setup'
        /home/alex/.local/share/nvim/nvt-dev.lua:34: in function 'setup'
        [string ":lua"]:1: in main chunk
        [C]: in function 'cmd'
        ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:359: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
stack traceback:
        [C]: in function 'cmd'
        ...vt-dev/site/pack/packer/start/packer.nvim/lua/packer.lua:359: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

Happens for git and diagnostics too.

Proposal:
- Only call `vim.fn.sign_define` if the user has specified `"signcolumn"`
- Trim the glyph to 2 characters

chomosuke commented:
@alex-courtis Sadly, because lua dose not do utf8 string properly. We can't determine the length of a string or trim a string without some dependency such as https://github.com/uga-rosa/utf8.nvim. I've also tried to pass the string via `nvim_eval` to vim script and trim it there but had no luck.

alex-courtis commented:
> @alex-courtis Sadly, because lua dose not do utf8 string properly. We can't determine the length of a string or trim a string without some dependency such as https://github.com/uga-rosa/utf8.nvim. I've also tried to pass the string via `nvim_eval` to vim script and trim it there but had no luck.

In that case, we will have to go with the first option:

> * Only call `vim.fn.sign_define` if the user has specified `"signcolumn"`

It is failing if you specify many characters even when `modified.enabled = false`

I really want to be able to use `[+]` `"after"` ;)

chomosuke commented:
OH YEAH sure, I forgot that option existed after trying so hard to do the second one haha.

alex-courtis commented:
The note is not enough...

Try running with all defaults and `glyphs.modified = "XXX"`

chomosuke commented:
whoops sorry, forgot to push.

alex-courtis commented:
awesome
```
    .hooks
    doc [+]
    lua
      nvim-tree
       nvim-tree.lua [+]
    scripts
```

alex-courtis reviewed (approved):
I will enjoy this.
